specs/fulu/beacon-chain.md

```diff
@@ -76,3 +88,156 @@ def process_execution_payload(state: BeaconState, body: BeaconBlockBody, executi
+        proposer_indices.append(compute_proposer_index(state, indices, seed))
+    return proposer_indices
+```
+
+#### New `compute_proposer_lookahead`
+
+```python
+def compute_proposer_lookahead(state: BeaconState) -> List[ValidatorIndex, (MIN_SEED_LOOKAHEAD + 1) * SLOTS_PER_EPOCH]:
+    """
+    Return the proposer indices for the full available lookahead starting from current epoch.
+    """
+    current_epoch = get_current_epoch(state)
+    lookahead = []
+    for i in range(MIN_SEED_LOOKAHEAD + 1):
+        proposer_indices = compute_proposer_indices(state, Epoch(current_epoch + i))
+        lookahead.extend(proposer_indices)
+    return lookahead        
+```
+
+
+### Beacon state accessors
+
+#### Modified `get_beacon_proposer_index`
+
+*Note*: The function `get_beacon_proposer_index` is modified to use the pre-calculated `current_proposer_lookahead` instead of calculating it on-demand.
+
+```python
+def get_beacon_proposer_index(state: BeaconState) -> ValidatorIndex:
+    """
+    Return the beacon proposer index at the current slot.
+    """
+    slot_offset = state.slot - compute_start_slot_at_epoch(get_current_epoch(state))
+    return state.proposer_lookahead[slot_offset]
+```
+
+### Epoch processing
+
+#### Modified `process_epoch`
+
+*Note*: The function `process_epoch` is modified in Fulu to call `process_proposer_lookahead` to update the `proposer_lookahead` in the beacon state.
+
+```python
+def process_epoch(state: BeaconState) -> None:
+    process_justification_and_finalization(state)
+    process_inactivity_updates(state)
+    process_rewards_and_penalties(state)
+    process_registry_updates(state)
+    process_slashings(state)
+    process_eth1_data_reset(state)
+    process_pending_deposits(state)
+    process_pending_consolidations(state)
+    process_effective_balance_updates(state)
+    process_slashings_reset(state)
+    process_randao_mixes_reset(state)
+    process_historical_summaries_update(state)
+    process_participation_flag_updates(state)
+    process_sync_committee_updates(state)
+    process_proposer_lookahead(state)  # [New in Fulu:EIPXXXX]
+```
+
+#### New `process_proposer_lookahead`
+
+*Note*: This function updates the proposer_lookahead in the beacon state by shifting out proposer indices from the earliest epoch and appending new proposer indices for the latest epoch. With `MIN_SEED_LOOKAHEAD` set to `1`, this means that at the start of epoch `N`, the proposer lookahead for epoch `N+1` will be computed and included in the beacon state's lookahead.
+
+```python
+def process_proposer_lookahead(state: BeaconState) -> None:
+    last_epoch_start = len(state.proposer_lookahead) - SLOTS_PER_EPOCH
+    # Shift out proposers in the first epoch.
+    state.proposer_lookahead[:last_epoch_start] = state.proposer_lookahead[SLOTS_PER_EPOCH:]
+    # Fill in the last epoch with new proposer indices.
+    last_epoch_proposers = compute_proposer_indices(state, Epoch(get_current_epoch(state) + MIN_SEED_LOOKAHEAD))
```

Comment:
Isn't this an off-by-one? `process_proposer_lookahead` is called in `process_epoch` which itself is called  in `process_slots` **before** the state slot is set, that is that `get_current_epoch(state)` will return the current epoch before the update and this will be computing the next's epoch proposers. I think here you want to be computing from one epoch more.

Reply:
Hmm, interesting.. What you are saying makes sense (good find!), but at the same time not sure why the tests are passing if this is the case. Let me take a deeper look tomorrow.

Reply:
You were completely right, thanks a lot for pointing out! It is fixed here: https://github.com/ethereum/consensus-specs/pull/4190/commits/208f3687d2ae7662c9de4805772dbb2eeaf2d1ca

Side note: My tests didn't catch this because the `yield from run_epoch_processing_with(..)` used in tests do NOT increment the slot number, hence the test and the specs were coincidentally both off by one :slightly_smiling_face: